core/client/hammer/hammer.go

```diff
@@ -76,125 +97,199 @@ func New(ctx context.Context, dial DialFunc, callOptions CallOptions,
-				refresh.Stop()
-				return
-			}
-		}
-	}()
+type worker struct {
+	*Hammer
+	client *client.Client
+	bar    *cpb.ProgressBar
+}
 
-	// Collect outputs until output closes.
-	for err := range outputs {
-		totalRequests++
+func (h *Hammer) newWorkers(n int) ([]worker, error) {
+	workers := make([]worker, 0, n)
+	for i := 0; i < n; i++ {
+		// Give each worker its own client.
+		client, err := client.NewFromConfig(h.ktCli, h.domain)
 		if err != nil {
-			log.Printf("writeOp(): %v", err)
+			return nil, err
 		}
+
+		workers = append(workers, worker{
+			Hammer: h,
+			client: client,
+			bar:    cpb.New(100).Prefix(fmt.Sprintf("%v ", i)),
+		})
 	}
-	close(stop)
-	wg.Wait()
+	return workers, nil
 }
 
-func (h *Hammer) worker(ctx context.Context, jobs <-chan int, outputs chan<- error) {
-	client, err := client.NewFromConfig(h.ktCli, h.config)
-	if err != nil {
-		glog.Errorf("NewFromConfig(): %v", err)
-		return
+// writeOp queues many user mutations, waits, and then verifies them all.
+func (w *worker) writeOp(ctx context.Context, req *opArg) error {
+	users := make([]*tpb.User, 0, len(req.UserIDs))
+	for _, userID := range req.UserIDs {
+		users = append(users, &tpb.User{
+			DomainId:       w.domain.DomainId,
+			AppId:          w.appID,
+			UserId:         userID,
+			PublicKeyData:  []byte("publickey"),
+			AuthorizedKeys: w.authorizedKeys,
+		})
+	}
+
+	mutations := make([]*entry.Mutation, 0, len(users))
+	for _, u := range users {
+		callOptions := w.callOptions(u.UserId)
+
+		cctx, cancel := context.WithTimeout(ctx, w.timeout)
+		m, err := w.client.CreateMutation(cctx, u)
+		cancel()
+		if err != nil {
+			return err
+		}
+		mutations = append(mutations, m)
+		cctx, cancel = context.WithTimeout(ctx, w.timeout)
+		err = w.client.QueueMutation(cctx, m, w.signers, callOptions...)
+		cancel()
+		if err != nil {
+			return err
+		}
 	}
 
-	for i := range jobs {
-		userID := fmt.Sprintf("user_%v", i)
-		outputs <- h.writeOp(ctx, client, userID)
+	for _, m := range mutations {
+		cctx, cancel := context.WithTimeout(ctx, w.timeout)
+		_, err := w.client.WaitForUserUpdate(cctx, m)
+		cancel()
+		if err != nil {
+			return err
+		}
+		fmt.Print(".")
 	}
+	return nil
 }
 
-// writeOp performs one write command.
-func (h *Hammer) writeOp(ctx context.Context, client *client.Client, userID string) error {
-	cctx, cancel := context.WithTimeout(ctx, h.timeout)
-	defer cancel()
-
-	u := &tpb.User{
-		DomainId:       h.config.DomainId,
-		AppId:          h.appID,
-		UserId:         userID,
-		PublicKeyData:  []byte("publickey"),
-		AuthorizedKeys: h.authorizedKeys,
+// readOp simulates multiple read operations by a single client.
+// Typical conversation setup involves querying two userIDs: self and other.
+func (w *worker) readOp(ctx context.Context, req *opArg) error {
+	for _, userID := range req.UserIDs {
+		_, _, err := w.client.GetEntry(ctx, userID, w.appID)
+		if err != nil {
```

Comment:
nit: Could merge this line with the previous one.

Reply:
done